### ISSUE TITLE
nestedErrors raises errors that are not errors

### ISSUE THREAD

**Issue by planttheidea** [tdegrunt/jsonschema#195]
**TL;DR**

When surfacing nested errors in cases of `oneOf` or `anyOf`, if a nested value of one of the types fails then all the other types in the array show an error.

I'm creating a schema that uses `oneOf`, and something like this:

```javascript
const VALIDATOR = new Validator();
const SCHEMA = {
  id: '/FooSchema',
  oneOf: [
    {
      properties: {
        foo: {
          type: 'number'
        }
      },
      type: 'object'
    }, {
      items: {
        type: 'string'
      },
      type: 'array'
    }
  ]
};
const OBJECT = {
  foo: null
};

const result = VALIDATOR.validate(OBJECT, SCHEMA, {nestedErrors: true});

result.errors.forEach((error) => {
  console.error(error.stack);
});
```

Will produce these error messages:

```javascript
instance.foo is not of type(s) number
instance is not of a type(s) array
instance is not exactly one from [subschema 0], [subschema 1]
```

The first and last one are definitely true, however the second one seems like it should not be surfacing. The nested error here the fact that the object with property foo fails validation, and bubbling up the fact it is a property on one of the valid types should satisfy the `oneOf` requirement and not show the instance type check error for the other type(s). This is true for all other types the instance does not match ... if I add a third type in the `oneOf` array, it will show an error for that too. This is also true if you switch to using `anyOf`.

At minimum this is noise, but it also serves as a point of confusion when diagnosing the source of the error.

**Comment by martinp999** [comment]
While I actually have a problem in the same area, I don't see that the example you gave is a problem.

Firstly, OBJECT is of type 'object' with a single property of type 'null', so it matches neither the first schema in your oneOf or the second one.

Secondly, as per the documentation - 

> When oneOf or anyOf validations fail, errors that caused any of the sub-schemas referenced therein to fail are not reported, **unless options.nestedErrors is truthy**

You have specifically set nestedErrors to true, so you're telling the validator to report them, which it is doing.

Or, have I missed something?

**Comment by martinp999** [comment]
The problem that I am having is that setting nestedErrors to **false** has no effect.

Due to the fact that I am running jsonschema in mocha, I need the validator to throw errors when it _properly fails_, which is why I am setting options.throwError to true.

The problem is that it is throwing errors even though it is _not properly failing_.  By that I mean - 

```
const VALIDATOR = new Validator();
const SCHEMA = {
  id: '/FooSchema',
  oneOf: [
    {
      properties: {
        foo: {
          type: 'number'
        }
      },
      type: 'object'
    }, {
      properties: {
        foo: {
          type: 'string'
        }
      },
      type: 'object'
    }
  ]
};
const OBJECT1 = {
  foo: 12
};
const OBJECT2 = {
  foo: '12'
};

VALIDATOR.validate(OBJECT1, SCHEMA, {throwError: true, nestedErrors: false});
VALIDATOR.validate(OBJECT2, SCHEMA, {throwError: true, nestedErrors: false});

```
both OBJECT1 and OBJECT2 will throw an error, even though I have nestedErrors set to false (in fact it throws an error no matter what I do with nestedErrors - true, false, or not handed in).  OBJECT1 will fail on the second schema I have in my oneOf (because the validator needs to make sure that the first schema is the _only_ one it matches), OBJECT2 will fail on the first.

I get it that both objects need to be validated against each of the schemas, but individual fails within oneOf should be swallowed and an error thrown only when _none_ of the schemas match.

Granted, nestedErrors seems to also have a problem - possibly due to a conflict with throwErrors, but, firstly, I think that the whole complexity of nestedErrors should not be necessary if oneOf properly absorbed internal mismatches, secondly, even if nestedErrors worked as advertised, it would not produce the correct result - I'm assuming that, if I could turn off nestedErrors, I would then never get a fail on oneOf, even when it matched none of the schemas.
